File: iOS_Sign/iOS_Sign/SignView/SignView.m
//
//  SignView.m
//  iOS_Sign
//
//  Created by Ryan on 2019/2/28.
//  Copyright © 2019 Ryan. All rights reserved.
//

#import "SignView.h"

@interface SignView () {
    CGPoint points[5];
}

@property (assign, nonatomic) NSInteger pointIndex;
@property (strong, nonatomic) UIBezierPath *bezierPath;

@end

@implementation SignView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    self.backgroundColor = [UIColor clearColor];
    [self setMultipleTouchEnabled:NO];
    
    _textColor = [UIColor redColor];
    _lineWidth = 3.f;
    _bezierPath = [UIBezierPath bezierPath];
    [_bezierPath setLineWidth:_lineWidth];
}

- (void)setLineWidth:(CGFloat)lineWidth {
    _lineWidth = lineWidth;
    [_bezierPath setLineWidth:_lineWidth];
}

#pragma mark -- core
- (void)drawRect:(CGRect)rect {
    [_textColor setStroke];
    [_bezierPath stroke];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    _pointIndex = 0;
    UITouch *touch = touches.anyObject;
    points[_pointIndex] = [touch locationInView:self];
    [_bezierPath moveToPoint:points[0]];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    _pointIndex++;
    UITouch *touch = touches.anyObject;
    points[_pointIndex] = [touch locationInView:self];
    if (_pointIndex == 4) {
        points[3] = CGPointMake((points[2].x + points[4].x)/2, (points[2].y + points[4].y)/2);
        [_bezierPath moveToPoint:points[0]];
        [_bezierPath addCurveToPoint:points[3] controlPoint1:points[1] controlPoint2:points[2]];
        [self setNeedsDisplay];
        
        //reset new start to last end
        points[0] = points[3];
        points[1] = points[4];
        _pointIndex = 1;
    }
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (_pointIndex < 4) {
        _pointIndex++;
        UITouch *touch = touches.anyObject;
        points[_pointIndex] = [touch locationInView:self];
        CGPoint to = points[_pointIndex];
        CGPoint center = CGPointZero;
        CGPoint from = CGPointZero;
        if (_pointIndex > 1) {
            center = points[_pointIndex-1];
            from = points[_pointIndex-2];
        } else {
            from = points[_pointIndex-1];
        }
        if (CGPointEqualToPoint(to, from)) {
            [_bezierPath addQuadCurveToPoint:CGPointMake(to.x + 2, to.y + 2) controlPoint:from];
            //圆点
//            [_bezierPath addArcWithCenter:from radius:1.f startAngle:0 endAngle:M_PI*2 clockwise:YES];
        } else if (CGPointEqualToPoint(center, CGPointZero)) {
            [_bezierPath addQuadCurveToPoint:to controlPoint:from];
        } else {
            [_bezierPath addCurveToPoint:to controlPoint1:from controlPoint2:center];
        }
        [self setNeedsDisplay];
    }
}


#pragma mark -- method
- (void)clear {
    [self.bezierPath removeAllPoints];
    [self setNeedsDisplay];
}

- (UIImage *)getSignImage {
    UIGraphicsBeginImageContextWithOptions(self.bounds.size, NO, [UIScreen mainScreen].scale);
    [self.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *signImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return signImage;
}

@end

File: iOS_Sign_Swift/iOS_Sign_Swift/SignView/SignView.swift
//
//  SignView.swift
//  iOS_Sign_Swift
//
//  Created by Ryan on 2019/2/28.
//  Copyright © 2019 Ryan. All rights reserved.
//

import UIKit

class SignView: UIView {

    public var textColor: UIColor = UIColor.red
    
    public var _lineWidth: CGFloat = 3.0
    var lineWidth: CGFloat{
        set {
            _lineWidth = newValue
            self.bezierPath.lineWidth = _lineWidth
        }
        get {
            return _lineWidth
        }
    }
    
    private var points: [CGPoint] = Array.init(repeating: CGPoint.zero, count: 5)
    private var pointIndex = 0
    private var bezierPath: UIBezierPath = UIBezierPath.init()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    private func commonInit() {
        self.backgroundColor = UIColor.clear
        self.isMultipleTouchEnabled = false
        bezierPath.lineWidth = 3.0
    }
    
    override func draw(_ rect: CGRect) {
        textColor.setStroke()
        bezierPath.stroke()
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        pointIndex = 0
        points[pointIndex] = touches.first!.location(in: self)
        bezierPath.move(to: points[0])
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        pointIndex = pointIndex + 1
        points[pointIndex] = touches.first!.location(in: self)
        if pointIndex == 4 {
            points[3] = CGPoint.init(x: (points[2].x + points[4].x)/2, y: (points[2].y + points[4].y)/2)
            bezierPath.move(to: points[0])
            bezierPath.addCurve(to: points[3], controlPoint1: points[1], controlPoint2: points[2])
            self.setNeedsDisplay()
            
            //reset new start to last end
            points[0] = points[3];
            points[1] = points[4];
            pointIndex = 1;
        }
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        if pointIndex < 4 {
            pointIndex = pointIndex + 1
            points[pointIndex] = touches.first!.location(in: self)
            let to: CGPoint = points[pointIndex]
            var center: CGPoint?
            var from: CGPoint?
            if pointIndex > 1 {
                center = points[pointIndex - 1]
                from = points[pointIndex - 2]
            } else {
                from = points[pointIndex - 1]
            }
            if to.equalTo(from!) {
                bezierPath.addQuadCurve(to: CGPoint.init(x: to.x + 2, y: to.y + 2), controlPoint: from!)
            } else if center != nil {
                bezierPath.addCurve(to: to, controlPoint1: from!, controlPoint2: center!)
            } else {
                bezierPath.addQuadCurve(to: to, controlPoint: from!)
            }
            self.setNeedsDisplay()
        }
    }
    
    public func clear() {
        self.bezierPath.removeAllPoints()
        self.setNeedsDisplay()
    }

    public func getSignImage() -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(self.bounds.size, false, UIScreen.main.scale)
        self.layer.render(in: UIGraphicsGetCurrentContext()!)
        if let signImage = UIGraphicsGetImageFromCurrentImageContext() {
            UIGraphicsEndImageContext()
            return signImage
        }
        return nil
    }
}
